README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# fuse.prep

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
[![Codecov test
coverage](https://codecov.io/gh/lhmet-ped/fuse.prep/branch/master/graph/badge.svg)](https://codecov.io/gh/lhmet-ped/fuse.prep?branch=master)
<!-- badges: end -->

O pacote **`{fuse.prep}`** tem o objetivo de gerar os dados de entrada
para aplicação do *Framework for Understanding Structural Errors*
([FUSE](https://naddor.github.io/fuse/)) nomodo de escala de bacia
hidrográfica. São necessários dois arquivos NetCDF de entrada para o
FUSE:

  - bandas de elevação do terreno: armazena as frações de área da bacia
    hidrográfica e da precipitação anual por banda de elevação;

  - forçantes meteorológicas: armazena as séries temporais da média
    diárias na área da bacia hidrográfica das variáveis: temperatura do
    ar (°C), precipitação (mm dia<sup>-1</sup>), evapotranspiração
    potencial (mm dia<sup>-1</sup>) e opcionalmente de vazão observada
    (ou deflúvio, mm dia<sup>-1</sup>);

A seguir apresenta-se como gerar estes dois arquivos a partir de dados
pré-processados fornecidos com os pacotes. Após conhecer estas funções,
você pode estar interessado em gerar os dados para sua bacia
hidrográfica de interesse. Este procedimento é detalhado nas vinhetas
fornecidas com o **`{fuse.prep}`**. Recomenda-se reproduzi-las na
seguinte ordem:

1.  `vignette('pp-elevbands', package = "fuse.prep")`

2.  `vignette('pp-forcmets', package = "fuse.prep")`

## Pré-requisitos do sistema

Como o **{`fuse.prep`}** depende dos pacotes **{`lhmetools`}**,
**{`HEgis`}** e **{`HEobs`}**. Para sistemas linux, você precisa
instalar:

  - unrar (\>= 5.61, instruções de instalação
    [aqui](https://github.com/lhmet/lhmetools/#system-requirements))

  - netcdf (\>= 4.7.3) e udunits-2 (instruções de instalação
    [aqui](https://github.com/ropensci/tidync#ubuntudebian))

  - GDAL (\>= 2.0.1), GEOS (\>= 3.4.0) and Proj.4 (\>= 4.8.0)
    (instruções de instalação
    [aqui](https://github.com/r-spatial/sf/blob/master/README.md#linux)).

No Windows você precisa do [7-zip](https://www.7-zip.org/), um software
livre, facilmente instalado a partir do R com o pacote **{`installr`}**
usando `installr::install.7zip()`.

## Instalação

O **{`fuse.prep`}** pode ser instalado do [GitHub](https://github.com/)
com:

``` r
# install.packages("devtools")
devtools::install_github(
  "lhmet-ped/fuse.prep",
  build_vignettes = TRUE,
  dependencies = TRUE
)
```

A opção `build_vignettes = TRUE` instalará o pacote incluindo as
vinhetas do pacote que apresentam as etapas de pré-processamento dos
dados necessários para criação dos arquivos NetCDF.

### NetCDF de bandas de elevação

Os dados de exemplo são da Bacia Hidrográfica associada ao posto 74 (UHE
G.B. MUNHOZ). Os dados necessários para geração do arquivo NetCDF de
bandas de elevação:

  - raster da precipitação climatológica anual

  - raster da elevação do terreno hidrologicamente condicionado

  - polígono da bacia hidrográfica (*Simple Feature*, `sf`)

Estes dados são disponibilizados com este pacote (`precclim74`) e com o
pacote **`{HEgis}`** (`poly74` e `condem74`).

``` r
library(HEgis)
poly74
#> Simple feature collection with 1 feature and 9 fields
#> geometry type:  POLYGON
#> dimension:      XY
#> bbox:           xmin: -51.72274 ymin: -26.85503 xmax: -48.94853 ymax: -25.22428
#> geographic CRS: SIRGAS 2000
#>   codONS codANA       nome                                   nomeOri    adkm2
#> 1     74   7659 G_B_MUNHOZ UHE Governador Bento Munhoz da Rocha Neto 30207.57
#>   volhm3        rio cobacia        tpopera                       geometry
#> 1   5779 Rio Iguaçu 8625591 Regulariza_ONS POLYGON ((-51.56304 -26.259...
condem74
#> class      : RasterLayer 
#> dimensions : 1957, 3329, 6514853  (nrow, ncol, ncell)
#> resolution : 0.0008333333, 0.0008333333  (x, y)
#> extent     : -51.7225, -48.94833, -26.855, -25.22417  (xmin, xmax, ymin, ymax)
#> crs        : +proj=longlat +datum=WGS84 +no_defs 
#> source     : memory
#> names      : layer 
#> values     : 588, 1506  (min, max)
library(fuse.prep)
precclim74
#> class      : RasterLayer 
#> dimensions : 8, 13, 104  (nrow, ncol, ncell)
#> resolution : 0.25, 0.25  (x, y)
#> extent     : -52, -48.75, -27, -25  (xmin, xmax, ymin, ymax)
#> crs        : +proj=longlat +datum=WGS84 +no_defs 
#> source     : memory
#> names      : layer 
#> values     : 1444.804, 2400.077  (min, max)
```

Para saber como estes 3 arquivos foram gerados consulte a vinheta de
*Pré-processamento dos dados de bandas de elevação* digitando
`vignette('pp-elevbands', package = "fuse.prep")`.

O arquivo NetCDF de bandas de elevação é gerado com a função
`elev_bands_ncname()` que requer como entrada os dados carregados acima
e o centróide do polígono. O centróide do polígono da bacia hidrográfica
é obtido com a função `centroids()`.

``` r
(poly_ctrd <- centroids(poly_station = poly74))
#> # A tibble: 1 x 3
#>     lon   lat station
#>   <dbl> <dbl> <chr>  
#> 1 -50.3 -26.0 74
```

Para gerar o arquivo precisamos definir onde ele será salvo.

``` r
# arquivo de saída - altere o caminho para o local desejado
# nome do arquivo netcdf: 'suffix_elev_bands.nc'
(suffix <- paste0("posto", poly74$codONS))
#> [1] "posto74"
dest_dir <- tempdir()
elev_bands_ncname <- file.path(
  dest_dir, 
  paste(suffix, "elev_bands.nc", sep = "_")
)
```

Agora temos os dados de entrada para função `elev_bands_ncname()` gerar
o arquivo NetCDF.

``` r
elev_bands_file <- elev_bands_nc(
  con_dem = condem74, 
  meteo_raster = precclim74, 
  nbands = 14,
  ccoords = poly_ctrd,
  file_nc = elev_bands_ncname,
  na = -9999
)
#>   |                                                                              |                                                                      |   0%  |                                                                              |==================                                                    |  25%  |                                                                              |===================================                                   |  50%  |                                                                              |====================================================                  |  75%  |                                                                              |======================================================================| 100%
#> 
#>   |                                                                              |                                                                      |   0%  |                                                                              |===================================                                   |  50%
elev_bands_file
#> [1] "/tmp/Rtmp5UUq7h/posto74_elev_bands.nc"
file.exists(elev_bands_file)
#> [1] TRUE
```

## NetCDF das forçantes meteorológicas

Para gerar o arquivo NetCDF com as séries temporais das forçantes
hidrometeorológicas médias na área da BH, utilizaremos o conjunto de
dados `forcdata74` disponibilizado com este pacote.

``` r
str(forcdata74)
#> tibble [13,149 × 6] (S3: tbl_df/tbl/data.frame)
#>  $ date   : Date[1:13149], format: "1980-01-01" "1980-01-02" ...
#>  $ station: num [1:13149] 74 74 74 74 74 74 74 74 74 74 ...
#>  $ temp   : num [1:13149] 20.6 20.6 20.6 20.6 20.6 ...
#>  $ pr     : num [1:13149] 0.3635 0.2136 0.169 0.226 0.0616 ...
#>  $ pet    : num [1:13149] 5.37 5.01 5.38 5.53 5.51 ...
#>  $ q_obs  : num [1:13149] 1.81 1.64 1.43 1.34 1.22 ...
```

Para saber como gerar estes dados consulte a vinheta de
*Pré-processamento dos dados das forçantes meteorológicas* digitando
`vignette('pp-forcmets', package = "fuse.prep")`.

Cada variável é passada como um vetor conforme código abaixo. Os nomes
dos vetores devem seguir este padrão: `pr` para precipitação, `pet` para
evapotranspiração potencial e `q_obs` para vazão observada. Além das
séries temporais das variáveis precisamos das datas, as coordenadas da
bacia hidrográfica do posto e o arquivo NetCDF que será salvo.

``` r
# arquivo de saída - altere o caminho para o local desejado
forcings_ncname <- file.path(
  dest_dir, 
  paste(suffix, "input.nc", sep = "_")
)
# exporta dados para netcdf
meteo_forcing_nc(
  forc_tbl = forcdata74,
  ccoords = centroids(poly_station = poly74),
  file_nc = forcings_ncname 
)
#> [1] "/tmp/Rtmp5UUq7h/posto74_input.nc"
file.exists(forcings_ncname)
#> [1] TRUE
```


NEWS.md
- [ ] incluir testes para `spatial_average`, `annual_climatology`, `annual_summary`

# fuse.prep 0.3.0

- [x] fix https://github.com/lhmet-ped/fuse.prep/issues/4#issuecomment-736991500

# fuse.prep 0.2.0

- [x] `comb_data` recebe um tibble com as forçantes meteorológicas, ao invés
de receber vetores de cada variável.

# fuse.prep 0.1.5

- [x] resolvida issue #4  (graças à @nelsonvnperu e @andreza-santos). 

# fuse.prep 0.0.4.91

- [x] incluir vinhetas de pré-processamento de dados das forçantes meteorológicas


# fuse.prep 0.0.4

- [x] adicionar `join_netcdfs.R` ao `data-raw`

- [x] `annual_climatology` tem um novo argumento `cutoff` para excluir valores
menores ou iguais a este limiar.

# fuse.prep 0.0.3

- [x] documentar função `meteo_forcing_nc()`

- [x] incluir testing da função `meteo_forcing_nc()`

# fuse.prep 0.0.2

- [x] Adicionar função para gerar arquivo de NetCDF das forçantes meteorológicas.

# fuse.prep 0.0.1

- [x] Adicionada função para gerar arquivo de NetCDF de bandas de elevação 
(`elev_bands_nc`).


R/meteo-forcings.R
#-------------------------------------------------------------------------------
#' Join hydrometeorological data for FUSE and convert streamflow from cumecs
#' to mm/day.
#'
#' @param prec a \code{\link[tibble]{tibble}} with daily precipitation (mm day^-1^)
#' @param et0 a \code{\link[tibble]{tibble}} with daily reference evapotranspiration (mm day^-1^)
#' @param qobs a \code{\link[tibble]{tibble}} with daily natural streamflow (cumecs)
#' @param area area (in square kilometres) that flow volume is averaged
#' over.
#' @param save logical, TRUE for export output data to a RDS file.
#' @param dest_dir a character with the name of where the data is to be
#' saved.
#' @param prefix character, prefiz for the output file name.
#' @return a \code{\link[tibble]{tibble}}, column q_obs is the `qnat` variable
#'  (input) expressed in mm (discharge).
#' @export
#'
comb_data <- function(prec, et0, qobs, area,
                      save = TRUE,
                      prefix = "hydrodata-posto-",
                      dest_dir = "output"
){
  checkmate::assert_choice("qnat", names(qobs))
 # prec = prec_posto; et0 = pet_posto; qobs = qobs_posto; area = area_posto; stn_id = 74
  hydrodata <- prec %>%
    dplyr::inner_join(et0, by = c("date", "posto")) %>%
    dplyr::inner_join(qobs, by = c("date", "posto"))


  hydrodata <- hydrodata %>%
  dplyr::mutate(
    q_obs = convert_flow(
      # mudar no HEobs o nome qnat para q_obs
      qnat,
      from = "m^3/sec",
      to = "mm/day",
      area.km2 = area[1]
    ),
    qnat = NULL
  ) %>%
    dplyr::rename("station" = posto)
  message("q_obs has been converted to mm.")

  if(save){
    hydrodata_file <- save_data(hydrodata,
                                .prefix = prefix,
                                .posto_id = hydrodata[["station"]][1],
                                .dest_dir = dest_dir
    )
    message(hydrodata_file)
  }
  hydrodata
}



#-------------------------------------------------------------------------------
#' Check inputs
#' @noRd
#' @family forcings functions
.check_inputs_meteo_forc <- function(variab_list, ctrd, file) {
  # variab_list = meteo_data
  checkmate::assert_choice("date", names(variab_list))
  checkmate::assert_choice("station", names(variab_list))

  # check for valid data
  meteo_data <- dplyr::select(variab_list, -date, -station)
  met_vnames <- names(meteo_data)
  all_obs_miss <- apply(meteo_data, 2, function(x) all(is.na(x)))
  checkmate::assert_true(sum(all_obs_miss) == 0)

  checkmate::assert_subset(met_vnames, all_variables())

  #identical_lengths <- all(diff(unname(unlist(lapply(variab_list, length)))) == 0)
  #checkmate::assert_true(identical_lengths)
  checkmate::assert_subset(class(variab_list$date), c("Date", "POSIXct", "POSIXt"))
  checkmate::assert_class(ctrd, "data.frame")
  checkmate::assert_subset(c("lon", "lat", "station"), names(ctrd))
  checkmate::assert_directory_exists(dirname(file))
  return(invisible(NULL))
}

#' Set names of all variables can save in NetCDF
#' @noRd
#' @family forcings functions
all_variables <- function() c("temp", "pr", "pet", "q_obs")


#' Set dimensions attributes tibble
#' @noRd
#' @family forcings functions
dim_atts_tbl <- function(cc, dts) {
  tibble::tibble(
    name = c("longitude", "latitude", "time"),
    units = c("degreesE", "degreesN", "days since 1970-01-01"),
    vals = list(
      cc[["lon"]],
      cc[["lat"]],
      as.numeric(dts)
    )
  )
}

#' Set variables attributes tibble
#' @noRd
#' @family forcings functions
#'
vars_atts_tbl <- function(vnames, dim_atts_list, na_value) {
  spatial_mode <- "Catchment"

  # define variables
  long_names <- c(
    paste0(spatial_mode, "-averaged daily temperature"),
    paste0(spatial_mode, "-averaged daily precipitation"),
    paste0(spatial_mode, "-averaged daily potential evapotranspiration"),
    "Daily discharge"
  )
  # order matters
  names(long_names) <- all_variables()

  var_units <- c("degC", rep("mm/day", 3))
  names(var_units) <- names(long_names)

  tibble::tibble(
    name = vnames,
    units = var_units[names(var_units) %in% vnames],
    dim = lapply(seq_along(name), function(i) dim_atts_list),
    missval = rep(na_value, length(name)),
    longname = long_names[names(long_names) %in% vnames]
  )
}


#' Set global attributes tibble
#' @noRd
#' @family forcings functions
#'
glob_atts_tbl <- function(nc_obj, id) {
  tibble::tibble(
    nc = nc_obj,
    varid = rep(0, 3),
    attname = c("title", "instituition", "history"),
    attval = c(
      paste0("FUSE forcing file for catchment ", id),
      "LHMET-UFSM",
      paste0(Sys.info()[["user"]], ": ", Sys.Date())
    )
  )
}


#' Create NetCDF file of Meteorological forcings
#'
#' @param forc_tbl tibble with time series of meteorological forcings.
#' @inheritParams elev_bands_nc
#' @export
#' @examples
#' if(FALSE){
#'  # arquivo de saída
#'  forcings_nc <- "inst/extdata/posto74_input.nc"
#'  # exporta dados para netcdf
#'  meteo_forcing_nc(
#'    forc_tbl = forcdata74,
#'    ccoords = centroids(poly_station = poly74),
#'    file_nc = forcings_nc
#'  )
#'  file.exists(forcings_nc)
#'}
#' @family forcings functions
meteo_forcing_nc <- function(forc_tbl,
                             ccoords,
                             file_nc = "inst/extdata/74_input.nc",
                             na = -9999,
                             force_v4 = TRUE) {
  # forc_tbl = meteo_data
  # ccoords = centroids(poly_station = poly74); na = -9999; file_nc = "inst/extdata/74_input.nc"; force_v4 = TRUE
  var_names <- all_variables()[all_variables() %in% names(forc_tbl)]

  # check inputs
  .check_inputs_meteo_forc(
    variab_list = forc_tbl,
    ctrd = ccoords,
    file = file_nc
  )

  dates <- forc_tbl$date

  # define dimensions
  dim_atts_l <- dim_atts_tbl(ccoords, dates) %>%
    purrr::pmap(., ncdf4::ncdim_def)

  # define variables attributes
  vars_atts_l <- vars_atts_tbl(var_names, dim_atts_l, na) %>%
    purrr::pmap(ncdf4::ncvar_def)

  # open nc
  nc_conn <- ncdf4::nc_create(
    filename = file_nc,
    vars = vars_atts_l,
    force_v4 #,verbose = TRUE
  )

  # write global atttributes
  glob_atts_l <- glob_atts_tbl(list(nc_conn), ccoords[["id"]]) %>%
    purrr::pmap(., ncdf4::ncatt_put)

  # write variables to file
  invisible(
    lapply(
      var_names,
      function(ivar) {
        # ivar = "temp"
        ncdf4::ncvar_put(
          nc = nc_conn,
          varid = .select_attr_var(vars_atts_l, ivar),
          vals = forc_tbl[[ivar]]
        )
      }
    )
  )

  ncdf4::nc_close(nc_conn)
  checkmate::assert_file_exists(file_nc)
  file_nc
}

# !TESTAR
# meteo_forcing_nc(forcdata74,  ccoords = centroids(poly_station = poly74))
